tensorboard/data/dispatching_provider_test.py

```diff
@@ -0,0 +1,337 @@
+            providers
+        )
+
+    def test_data_location(self):
+        self.assertEqual(
+            self.with_unpfx.data_location(_ctx(), experiment_id="foo:123"),
+            "foo://123",
+        )
+        self.assertEqual(
+            self.with_unpfx.data_location(_ctx(), experiment_id="bar:a:b:c"),
+            "Bar://a:b:c",
+        )
+        self.assertEqual(
+            self.with_unpfx.data_location(_ctx(), experiment_id="baz"),
+            "BAZ://baz",
+        )
+        with self.assertRaisesRegex(
+            errors.NotFoundError, "Unknown data provider key: 'quux'"
+        ):
+            self.with_unpfx.data_location(_ctx(), experiment_id="quux:hmm")
+        with self.assertRaisesRegex(
+            errors.NotFoundError, "No unprefixed data provider specified"
+        ):
+            self.without_unpfx.data_location(_ctx(), experiment_id="quux")
+
+    def test_scalars(self):
+        listing = self.with_unpfx.list_scalars(
+            _ctx(), experiment_id="foo:123", plugin_name="scalars"
+        )
+        self.assertEqual(
+            listing,
+            {
+                "123/train": {
+                    "loss.scalars": provider.ScalarTimeSeries(
+                        max_step=2,
+                        max_wall_time=0.5,
+                        plugin_content=b"",
+                        description="Hello from foo",
+                        display_name="loss",
+                    )
+                }
+            },
+        )
+
+        reading = self.with_unpfx.read_scalars(
+            _ctx(),
+            experiment_id="foo:123",
+            plugin_name="scalars",
+            downsample=1000,
+            run_tag_filter=provider.RunTagFilter(
+                ["123/train"], ["loss.scalars"]
+            ),
+        )
+        self.assertEqual(
+            reading,
+            {
+                "123/train": {
+                    "loss.scalars": [
+                        provider.ScalarDatum(
+                            step=0, wall_time=0.0, value=float(len("scalars"))
+                        ),
+                        provider.ScalarDatum(
+                            step=1, wall_time=0.5, value=float(len("123"))
+                        ),
+                    ]
+                }
+            },
+        )
+
+    def test_blob_sequences(self):
+        def get_blobs(dp, experiment_id):
+            """List, read, and fetch all blobs in a given context."""
+            listing = dp.list_blob_sequences(
+                _ctx(), experiment_id=experiment_id, plugin_name="images"
+            )
+            reading = dp.read_blob_sequences(
+                _ctx(),
+                experiment_id=experiment_id,
+                plugin_name="images",
+                downsample=10,
+                run_tag_filter=provider.RunTagFilter(
+                    runs=list(listing),
+                    tags=sorted(set().union(*listing.values())),
```

Comment:
This feels a wee bit fancy for a test helper?  I can parse it, but it might be simpler to just make run and tag parameters to the helper method [1] or to leave then empty to request all runs and tags.  See also other comments but I'd probably try to keep the test logic here fairly focused on the parts that DispatchingProvider does differently versus things that are more like generic DataProvider behavior expectations.

[1] Another option rather than making them parameters is just to down-scope the helper to postprocessing the `read_blob_sequences()` response and making the `read_blob()` calls, since that's where I think it has the most bang for the buck, and just leaving the `{list,read}_blob_sequences()` calls as being directly invoked in the test method.